utilities/env_librados.cc

```diff
@@ -0,0 +1,961 @@
+  }
+};
+
+
+// A file abstraction for sequential writing.  The implementation
+// must provide buffering since callers may append small fragments
+// at a time to the file.
+class LibradosWritableFile : public rocksdb::WritableFile {
+    librados::IoCtx * _io_ctx;
+    std::string _fid;
+    const int BUFFER_MAX_SIZE = 1 << 12;
+    librados::bufferlist buffer;
+    int buffer_size;
+ public:
+    LibradosWritableFile(librados::IoCtx * io_ctx, std::string fid):
+        _io_ctx(io_ctx), _fid(fid), buffer(), buffer_size(0){
+    }
+    ~LibradosWritableFile() {
+        Sync();
+    }
+
+  rocksdb::Status Append(const rocksdb::Slice& data) {
+    LOG_DEBUG("%s|%i\n", _fid.c_str(), (int)data.size());
+
+    buffer.append(data.data());
+    buffer_size += data.size();
+
+    if (buffer_size > BUFFER_MAX_SIZE) {
+        return Flush();
+    } else {
+        return rocksdb::Status::OK();
+    }
+#if 0
+    librados::bufferlist buf;
+    buf.append(data.data());
+    int r = _io_ctx->append(_fid, buf, data.size());
+    return err_to_status(r);
+#endif
+  }
+
+  // Positioned write for unbuffered access default forward
+  // to simple append as most of the tests are buffered by default
+  rocksdb::Status PositionedAppend(
+    const rocksdb::Slice& /* data */,
+    uint64_t /* offset */) {
+    return rocksdb::Status::NotSupported();
+  }
+
+  // Truncate is necessary to trim the file to the correct size
+  // before closing. It is not always possible to keep track of the file
+  // size due to whole pages writes. The behavior is undefined if called
+  // with other writes to follow.
+  rocksdb::Status Truncate(uint64_t size) {
+    _io_ctx->trunc(_fid, size);
+    return rocksdb::Status::OK();
+  }
+  rocksdb::Status Close() {
+    return Flush();
+  }
+
+  rocksdb::Status Flush() {
+    // TODO: 
+    // Flush could initiate an aio write and not wait, 
+    // and Sync could Flush and then wait for any pending writes
+    return Sync();
+  }
+
+  rocksdb::Status Sync() { // sync data
+    if (buffer_size != 0) {
+        int r = _io_ctx->append(_fid, buffer, buffer_size);
+        buffer_size = 0;
+        buffer.clear();
+        return err_to_status(r);
+    } else {
+        return rocksdb::Status::OK();
+    } 
+  }
+
+  // true if Sync() and Fsync() are safe to call concurrently with Append()
+  // and Flush().
+  bool IsSyncThreadSafe() const {
+    // TODO: 
+    // if true,  need a mutex around the position and/or bufferlist members
+    return false;
+  }
+
+  // Indicates the upper layers if the current WritableFile implementation
+  // uses direct IO.
+  bool UseDirectIO() const {
+    return false;
+  }
+
+  /*
+   * Get the size of valid data in the file.
+   */
+  uint64_t GetFileSize() {
+      uint64_t fsize = 0;
+      int ret = _io_ctx->stat(_fid, &fsize, nullptr);
+      LOG_DEBUG(" %s|%lld|%i\n", _fid.c_str(), (long long)fsize, ret);
+      if (ret < 0) {
+          return 0;
+      } else {
+          return fsize;
+      }
```

Comment:
I think this could be replaced with an uint64_t size member that is updated by append and truncate.